File: 22-GenerateParentheses/22-GenerateParentheses.java
// Last updated: 01/09/2025, 01:06:49
class Solution {
    public void generating(List<String>res,String s, int open, int close, int size){
        if(open == size && close == size){
            res.add(s);
            return;
        }
        if(open < size){
            generating(res, s + '(' , open + 1, close, size);
        }
        if(open > close){
            generating(res, s + ')' , open , close + 1, size);
        }
    }
    public List<String> generateParenthesis(int n) {
        List<String>lst = new ArrayList<>();
        generating(lst, "", 0, 0, n);
        return lst;
    }
}

File: 22-GenerateParentheses/22-GenerateParentheses.cpp
// Last updated: 01/09/2025, 01:06:52
class Solution {
public:
    void generatePara(int close, int open, string s, vector<string>&res, int n){
        if(close == open && (open + close) == 2 * n){
            res.push_back(s);
            return;
        }

        if(open < n){
            generatePara(close,open+1,s+'(',res,n);
        }

        if(close < open){
            generatePara(close+1,open,s+')',res,n);
        }

    }
    vector<string> generateParenthesis(int n) {
        vector<string>res;
        string s;
        generatePara(0,0,s,res,n);
        return res;
    }
};